src/Hl7.Fhir.Base/Serialization/PathStack.cs

```diff
@@ -43,33 +51,68 @@ public void ExitResource()
                 throw new InvalidOperationException("Cannot exit a resource while inside a property.");
             }
             else
+            {
                 _path = _path.Substring(0, top);
+            }
+
+            if (_paths.Count() == 1)
+                _paths.Pop();
         }
 
-        public void EnterElement(string name)
+        public void EnterElement(string name, int? index, bool isPrimitive)
         {
             bool empty = !_tops.Any();
             _tops.Push(_path.Length);
+            _indexer.Push(index ?? 0);
             _path += empty ? name : $".{name}";
+            if (!isPrimitive)
+            {
+                if (index.HasValue)
+                    _paths.Push($"{name}[{index}]");
+                else
+                    _paths.Push(name);
+            }
+            else
+            {
+                _paths.Push(String.Empty);
+            }
         }
 
         public void ExitElement()
         {
             if (_tops.Count == 0)
                 throw new InvalidOperationException("No resource or path parts are present on the stack.");
 
+
             var top = _tops.Pop();
+
             if (_path[top] == RESOURCEPREFIX)
             {
                 _tops.Push(top);
                 throw new InvalidOperationException("Cannot exit a property while inside a resource.");
             }
             else
+            {
+                _indexer.Pop();
                 _path = _path.Substring(0, top);
+            }
+
+            _paths.Pop();
+        }
+
+        public void IncrementIndex()
```

Comment:
It seems to me we now have the knowledge of indexed property names divided over two classes: the parser needs to pass the initial index + call `IncrementIndex`, and the PathStack needs to handle the effects.  Can't we design this so we have the parser call EnterElement, etc. without "knowing" anything about indexed paths? I don't like to introduce this "knowledge" into the parser, just to keep of the instance path.  Maybe the parser needs to pass a little bit more metadata into the calls, but so be it!